Source/com/drew/metadata/gif/GifReader.java

```diff
@@ -36,9 +56,75 @@
 
     public void extract(@NotNull final SequentialReader reader, final @NotNull Metadata metadata)
     {
-        GifHeaderDirectory directory = new GifHeaderDirectory();
-        metadata.addDirectory(directory);
+        reader.setMotorolaByteOrder(false);
+
+        GifHeaderDirectory header;
+        try {
+            header = ReadGifHeader(reader);
+            metadata.addDirectory(header);
+        } catch (IOException ex) {
+            metadata.addDirectory(new ErrorDirectory("IOException processing GIF data"));
+            return;
+        }
+
+        if(header.hasErrors())
+            return;
+
+        try {
+            // Skip over any global colour table
+            Integer globalColorTableSize = header.getInteger(GifHeaderDirectory.TAG_COLOR_TABLE_SIZE);
+            if (globalColorTableSize != null)
+            {
+                // Colour table has R/G/B byte triplets
+                reader.skip(3 * globalColorTableSize);
+            }
+
+            // After the header comes a sequence of blocks
+            while (true)
+            {
+                byte marker;
+                try {
+                    marker = reader.getInt8();
+                } catch (IOException ex) {
+                    return;
+                }
 
+                switch (marker)
+                {
+                    case (byte)'!': // 0x21
+                    {
+                        metadata.addDirectory(ReadGifExtensionBlock(reader));
+                        break;
+                    }
+                    case (byte)',': // 0x2c
+                    {
+                        metadata.addDirectory(ReadImageBlock(reader));
+
+                        // skip image data blocks
+                        SkipBlocks(reader);
+                        break;
+                    }
+                    case (byte)';': // 0x3b
+                    {
+                        // terminator
+                        return;
+                    }
+                    default:
+                    {
+                        // Anything other than these types is unexpected.
+                        // GIF87a spec says to keep reading until a separator is found.
+                        // GIF89a spec says file is corrupt.
+                        return;
+                    }
+                }
+            }
+        } catch (IOException e) {
+            metadata.addDirectory(new ErrorDirectory("IOException processing GIF data"));
+        }
+    }
+
+    private static GifHeaderDirectory ReadGifHeader(@NotNull final SequentialReader reader) throws IOException
```

Comment:
Method names in this class should have lowercase first letters to be more Javaesque.